trace/tests/unit/test_trace.py

```diff
@@ -0,0 +1,208 @@
+
+        return Trace
+
+    def _make_one(self, *args, **kw):
+        return self._get_target_class()(*args, **kw)
+
+    def test_constructor_defaults(self):
+        trace_id = 'test_trace_id'
+
+        patch = mock.patch(
+            'opencensus.trace.trace.generate_trace_id',
+            return_value=trace_id)
+
+        with patch:
+            trace = self._make_one()
+
+        self.assertEqual(trace.trace_id, trace_id)
+
+    def test_constructor_explicit(self):
+        trace_id = 'test_trace_id'
+        reporter = mock.Mock()
+
+        trace = self._make_one(
+            project_id=self.project,
+            trace_id=trace_id,
+            reporter=reporter)
+
+        self.assertEqual(trace.project_id, self.project)
+        self.assertEqual(trace.trace_id, trace_id)
+        self.assertIs(trace.reporter, reporter)
+
+    def test_start(self):
+        trace = self._make_one(project_id=self.project)
+        trace.start()
+
+        self.assertEqual(trace.spans, [])
+
+    def test_finish_with_valid_span(self):
+        from opencensus.trace.enums import TraceSpan as Enum
+        from opencensus.trace.trace_span import TraceSpan
+
+        reporter = mock.Mock()
+        trace = self._make_one(reporter=reporter)
+
+        span_name = 'span'
+        span_id = 123
+        kind = Enum.SpanKind.SPAN_KIND_UNSPECIFIED
+        start_time = '2017-06-25'
+        end_time = '2017-06-26'
+
+        span = mock.Mock(spec=TraceSpan)
+        span.name = span_name
+        span.kind = kind
+        span.parent_span_id = None
+        span.span_id = span_id
+        span.start_time = start_time
+        span.end_time = end_time
+        span.labels = None
+        span.children = []
+        span.__iter__ = mock.Mock(return_value=iter([span]))
+
+        with trace:
+            trace.spans = [span]
+            self.assertEqual(trace.spans, [span])
+
+        self.assertEqual(trace.spans, [])
+
+    def test_span(self):
+        from opencensus.trace.trace_span import TraceSpan
+
+        span_name = 'test_span_name'
+
+        trace = self._make_one(project_id=self.project)
+        trace.spans = []
+
+        trace.span(name=span_name)
+        self.assertEqual(len(trace.spans), 1)
+
+        result_span = trace.spans[0]
+        self.assertIsInstance(result_span, TraceSpan)
+        self.assertEqual(result_span.name, span_name)
+
+    def test_send_without_spans(self):
+        trace_id = 'test_trace_id'
+        reporter = mock.Mock()
+        trace = self._make_one(
+            project_id=self.project,
+            trace_id=trace_id,
+            reporter=reporter)
+        trace.spans = []
+
+        trace.send()
+
+        self.assertFalse(reporter.called)
+        self.assertEqual(trace.project_id, self.project)
+        self.assertEqual(trace.trace_id, trace_id)
+        self.assertEqual(trace.spans, [])
+
+    def test_send_with_spans(self):
+        from opencensus.trace.enums import TraceSpan as Enum
```

Comment:
I put those enum stuff all under a Enum class and removed the TraceSpan class under enum to avoid the confusion.